### PR TITLE
Add LinearFold

### PR DESCRIPTION
**Pull request by v-raja**
This PR adds a Go implementation of [LinearFold](https://github.com/LinearFold/LinearFold) to poly.

@Koeng101 would like to understand and add comments to the code so I've left almost all function, variable names, and comments as is from the original repo. To distinguish between my comments and comments from the original repo, I've prefixed all my comments with "vivek: ".

### CODE REVIEW

File: linearfold.go [hunk 1]
```diff
@@ -0,0 +1,1361 @@
...
+}
+
+func (s *State) Set2(score float64, manner int, l1 rune, l2 int) {
+	s.score = score
+	s.manner = manner
+	s.trace.paddings.l1 = l1
+	s.trace.paddings.l2 = l2
+}
+
+func (s *State) Set3(score float64, manner int, split int) {
```
Comment (TimothyStiles) [inline linearfold.go:339]: Why are there 3 set state methods and why are they are setting slightly different things that are unnamed?

Comment (Koeng101) [reply linearfold.go:339]: Yes.

Comment (TimothyStiles) [reply linearfold.go:339]: Spanish laughing guy collage

File: linearfold/linearfold.go [hunk 2]
```diff
@@ -0,0 +1,1230 @@
...
+	SET_HELIX_STACKING('U', 'A', 'A', 'U', true)
+	SET_HELIX_STACKING('U', 'A', 'G', 'U', true)
+	SET_HELIX_STACKING('U', 'G', 'G', 'U', true)
+}
+
+func SET_HELIX_STACKING(x rune, y rune, z rune, w rune, val bool) {
+	_helix_stacking[GET_ACGU_NUM(x)][GET_ACGU_NUM(y)][GET_ACGU_NUM(z)][GET_ACGU_NUM(w)] = val
+}
+
+func Min(a, b int) int {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:204]: why exported? Also duplicated from MRE PR.

File: linearfold/linearfold.go [hunk 3]
```diff
@@ -0,0 +1,1230 @@
...
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func Max(a, b int) int {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:211]: why exported?

File: linearfold/linearfold.go [hunk 4]
```diff
@@ -0,0 +1,1230 @@
...
+	if a > b {
+		return a
+	}
+	return b
+}
+
+func initialize_cachesingle() {
+	for i := 0; i < SINGLE_MAX_LEN+1; i++ {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:219]: golang variables don't have underscores

File: linearfold/linearfold.go [hunk 5]
```diff
@@ -0,0 +1,1230 @@
...
+				}
+			}
+
+		}
+	}
+}
+
+// What is manner?
+const (
+	MANNER_NONE                  = iota // 0: empty
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:265]: no underscores. uncapitalize. Variables are capitalized if exported.

File: linearfold/linearfold.go [hunk 6]
```diff
@@ -0,0 +1,1230 @@
...
+	MANNER_M_eq_M2                      // 9: M = M2
+	MANNER_M_eq_M_plus_U                // 10: M = M + U
+	MANNER_M_eq_P                       // 11: M = P
+	/* MANNER_C_eq_U, */
+	/* MANNER_C_eq_P, */
+	MANNER_C_eq_C_plus_U // 12: C = C + U
+	MANNER_C_eq_C_plus_P // 13: C = C + P
+)
+
+//
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:283]: State of what?

File: linearfold/linearfold.go [hunk 7]
```diff
@@ -0,0 +1,1230 @@
...
+			l2 int
+		}
+	}
+}
+
+func NewState() *State {
+	return &State{score: VALUE_MIN, manner: MANNER_NONE}
+}
+
+func update_if_better(s *State, newscore float64, manner int) {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:302]: Why 3? is there better terminology based on what these do?

File: linearfold/linearfold.go [hunk 8]
```diff
@@ -0,0 +1,1230 @@
...
+	if s.score < newscore {
+		s.Set(newscore, manner)
+	}
+}
+func update_if_better2(s *State, newscore float64, manner int, l1 rune, l2 int) {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:307]: again no underscores.

File: linearfold/linearfold.go [hunk 9]
```diff
@@ -0,0 +1,1230 @@
...
+	}
+}
+
+func update_if_better3(s *State, newscore float64, manner int, split int) {
+	if s.score < newscore || s.manner == MANNER_NONE {
+		s.Set3(newscore, manner, split)
+	}
+}
+
+func (s *State) Set(score float64, manner int) {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:319]: why 3? Can these be named better depending on context of what they do?

File: linearfold/linearfold.go [hunk 10]
```diff
@@ -0,0 +1,1230 @@
...
+
+func (s *State) Set3(score float64, manner int, split int) {
+	s.score = score
+	s.manner = manner
+	s.trace.split = split
+}
+
+// Assumes valid sequence input
+// Output: Linearfold score, structure
+func LinearFold(sequence string) (string, float64) {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:339]: can these return errors and catch invalid input?

File: linearfold/linearfold.go [hunk 11]
```diff
@@ -0,0 +1,1230 @@
...
+	return Parse(sequence)
+}
+
+type Pair struct {
+	first, second interface{}
+}
+
+type PairHeap []Pair
+
+func (h PairHeap) Len() int           { return len(h) }
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:360]: what are these one liners? Can they be more readable and use non single letter variables?

File: linearfold/linearfold.go [hunk 12]
```diff
@@ -0,0 +1,1230 @@
...
+
+func (h *PairHeap) Pop() interface{} {
+	old := *h
+	n := len(old)
+	x := old[n-1]
+	*h = old[0 : n-1]
+	return x
+}
+
+func Parse(sequence string) (string, float64) {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:378]: what are we parsing here?

File: linearfold/linearfold.go [hunk 13]
```diff
@@ -0,0 +1,1230 @@
...
+				// so no need to check the score
+
+				if bestH[nextComplement][i] == nil {
+					bestH[nextComplement][i] = NewState()
+				}
+				update_if_better(bestH[nextComplement][i], newscore, MANNER_H)
+			}
+		}
+
+		if j == 0 {
```
Comment (TimothyStiles) [inline linearfold/linearfold.go:571]: what is `j`?

**Review by TimothyStiles — COMMENTED** [review]
Awesome start. Variables to be in a Go friendly format. Single letter variables should also be renamed to something else. Helper functions need comments. Also go linter has seemed to catch a few print things that you probably want to fix.

**Comment by v-raja** [comment]
As per my discussion with @Koeng101, I've updated LinearFold to include the beam size variable (as this needs to be set to a small value for large sequences). (This allows me to continue my work on the RBS calculator)

**Comment by Koeng101** [comment]
Closed in favor of #273

**Pull request closed without merging**